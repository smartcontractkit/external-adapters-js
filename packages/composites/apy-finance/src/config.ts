import { util } from '@chainlink/ea-bootstrap'

export type Config = {
  rpcUrl: string
  registryAddr: string
}

export const makeConfig = (): Config => {
  return {
    rpcUrl: util.getRequiredEnv('RPC_URL'),
    registryAddr: util.getRequiredEnv('REGISTRY_ADDRESS'),
  }
}
