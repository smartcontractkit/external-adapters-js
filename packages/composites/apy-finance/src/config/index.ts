import { Requester, util } from '@chainlink/ea-bootstrap'
import * as types from '@chainlink/types'

export type Config = types.Config & {
  rpcUrl: string
  registryAddr: string
}

export const DEFAULT_ENDPOINT = 'tvl'
export const NAME = 'APY_FINANCE'

export const makeConfig = (prefix?: string): Config => {
  return {
    ...Requester.getDefaultConfig(prefix),
    rpcUrl: util.getRequiredEnvWithFallback('ETHEREUM_RPC_URL', ['RPC_URL'], prefix),
    registryAddr: util.getRequiredEnv('REGISTRY_ADDRESS', prefix),
    defaultEndpoint: DEFAULT_ENDPOINT,
  }
}
