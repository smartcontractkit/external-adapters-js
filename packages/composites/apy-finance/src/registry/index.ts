import { BigNumber, ethers } from 'ethers'
import registryAbi from '../abi/IRegistry.json'
import assetAllocationAbi from '../abi/IAssetAllocation.json'
import erc20 from '@openzeppelin/contracts/build/contracts/ERC20.json'
import { types } from '@chainlink/token-allocation-adapter'

export type GetAllocations = (
  registry: ethers.Contract,
  decimalsOf: (address: string) => Promise<number>,
) => () => Promise<types.TokenAllocations>

const getAllocations: GetAllocations = (registry, decimalsOf) => async () => {
  const tokenAddresses = await registry.getTokenAddresses()
  const [components, balances, decimals]: any = await Promise.all([
    Promise.all(tokenAddresses.map((address: string) => registry.symbolOf(address))),
    Promise.all(tokenAddresses.map((address: string) => registry.balanceOf(address))),
    Promise.all(tokenAddresses.map(async (address: string) => decimalsOf(address))),
  ])

  return components.map((symbol: string, i: number) => ({
    symbol,
    balance: BigNumber.from(balances[i]),
    decimals: decimals[i],
  }))
}

type Registry = {
  getAllocations: () => Promise<types.TokenAllocations>
}

const makeRegistry = async (address: string, rpcUrl: string): Promise<Registry> => {
  const provider = new ethers.providers.JsonRpcProvider(rpcUrl)
  const registry = new ethers.Contract(address, registryAbi, provider)
  const chainlinkRegistryAddress = await registry.chainlinkRegistryAddress()
  const chainlinkRegistry = new ethers.Contract(
    chainlinkRegistryAddress,
    assetAllocationAbi,
    provider,
  )

  const _getERC20 = (address: string) => new ethers.Contract(address, erc20.abi, provider)
  const _decimalsOf = (address: string) => _getERC20(address).decimals()

  return {
    getAllocations: getAllocations(chainlinkRegistry, _decimalsOf),
  }
}

export default makeRegistry
