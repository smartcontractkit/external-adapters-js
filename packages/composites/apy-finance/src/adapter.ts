import { Validator } from '@chainlink/external-adapter'
import { AdapterResponse, Execute, AdapterRequest } from '@chainlink/types'
import * as TokenAllocation from '@chainlink/token-allocation-adapter'
import makeRegistry from './registry'
import { makeConfig, Config } from './config'

export const execute = async (input: AdapterRequest, config: Config): Promise<AdapterResponse> => {
  const validator = new Validator(input)
  if (validator.error) throw validator.error

  const registry = await makeRegistry(config.registryAddr, config.rpcUrl)
  const allocations = await registry.getAllocations()

  const _execute = TokenAllocation.makeExecute()
  return await _execute({ id: validator.validated.id, data: { ...input.data, allocations } })
}

export const makeExecute = (config?: Config): Execute => {
  return async (request: AdapterRequest) => execute(request, config || makeConfig())
}
