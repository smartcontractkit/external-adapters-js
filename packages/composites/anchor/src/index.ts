import { expose } from '@chainlink/ea-bootstrap'
import { endpointSelector, makeExecute } from './adapter'
import { makeConfig, NAME } from './config'

/**
 * NOTE: This adapter could probably be made more general in the future.  All it is really doing is fetching an intermediary token price
 * using the Token Allocation adapter and handling the price after that.
 */

const adapterContext = { name: NAME }

const { server } = expose(adapterContext, makeExecute(), undefined, endpointSelector)

export { NAME, makeConfig, makeExecute, server }
