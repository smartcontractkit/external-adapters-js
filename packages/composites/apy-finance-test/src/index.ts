import { expose, ServerInstance } from '@chainlink/external-adapter-framework'
import { Adapter } from '@chainlink/external-adapter-framework/adapter'
import { allocations, tvl } from './endpoint'
import { config } from './config'

export const adapter = new Adapter({
  defaultEndpoint: tvl.name,
  name: 'APY_FINANCE_TEST',
  config,
  endpoints: [allocations, tvl],
})

export const server = (): Promise<ServerInstance | undefined> => expose(adapter)
