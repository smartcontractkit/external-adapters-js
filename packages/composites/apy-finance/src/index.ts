import { expose } from '@chainlink/ea-bootstrap'
import { makeExecute } from './adapter'
import { makeConfig } from './config'

const NAME = 'APY-Finance'

export = { NAME, makeConfig, makeExecute, ...expose(NAME, makeExecute()) }
