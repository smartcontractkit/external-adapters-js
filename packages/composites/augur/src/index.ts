import { expose } from '@chainlink/ea-bootstrap'
import { makeExecute } from './adapter'
import { makeConfig } from './config'

const NAME = 'AUGUR'

export = { NAME, makeConfig, makeExecute, ...expose(NAME, makeExecute()) }
