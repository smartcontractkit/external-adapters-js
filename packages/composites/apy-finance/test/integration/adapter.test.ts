import { AdapterRequest, Execute } from '@chainlink/types'
import nock from 'nock'
import * as apyFinanceAdapter from '../../src/index'
import { mockEthereumCalls, mockTiingoResponse } from './fixtures'
let oldEnv: NodeJS.ProcessEnv

describe('execute', () => {
  let execute: Execute
  const id = '1'

  beforeAll(() => {
    execute = apyFinanceAdapter.makeExecute()
    oldEnv = JSON.parse(JSON.stringify(process.env))
    process.env.CACHE_ENABLED = 'false'
    process.env.REGISTRY_ADDRESS =
      process.env.REGISTRY_ADDRESS || '0x7ec81b7035e91f8435bdeb2787dcbd51116ad303'
    process.env.TIINGO_DATA_PROVIDER_URL =
      process.env.TIINGO_DATA_PROVIDER_URL || 'http://localhost:3000'
    process.env.RPC_URL = process.env.RPC_URL || 'https://geth-main.eth.devnet.tools'
  })

  afterAll(() => {
    process.env = oldEnv
    if (process.env.RECORD) {
      nock.recorder.play()
    }

    nock.restore()
    nock.cleanAll()
    nock.enableNetConnect()
  })

  describe('with request properly formatted', () => {
    mockTiingoResponse()
    mockEthereumCalls()
    const data: AdapterRequest = {
      id,
      data: {
        source: 'tiingo',
      },
    }

    it.skip('should return success', async () => {
      const resp = await execute(data, {})
      expect(resp).toMatchSnapshot()
    })
  })
})
