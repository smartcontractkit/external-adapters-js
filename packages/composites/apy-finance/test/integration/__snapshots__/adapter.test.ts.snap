// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`execute with request properly formatted should return success 1`] = `
Object {
  "data": Object {
    "payload": Object {
      "USDC": Object {
        "quote": Object {
          "USD": Object {
            "price": 2000,
          },
        },
      },
    },
    "result": 200000,
    "sources": Array [],
  },
  "debug": undefined,
  "jobRunID": "1",
  "result": 200000,
  "statusCode": 200,
}
`;
